src/liballoc/collections/btree/navigate.rs

```diff
@@ -0,0 +1,116 @@
+use core::ptr;
+
+use super::node::{marker, ForceResult::*, Handle, NodeRef};
+
+impl<'a, K: 'a, V: 'a, NodeType, HandleType>
+    Handle<NodeRef<marker::Immut<'a>, K, V, NodeType>, HandleType>
+{
+    fn tree_hopper_copy(&self) -> Self {
+        *self
+    }
+}
+
+impl<'a, K: 'a, V: 'a, NodeType, HandleType>
+    Handle<NodeRef<marker::Mut<'a>, K, V, NodeType>, HandleType>
+{
+    unsafe fn tree_hopper_copy(&self) -> Self {
+        ptr::read(self)
+    }
+}
+
+impl<K, V, NodeType, HandleType> Handle<NodeRef<marker::Owned, K, V, NodeType>, HandleType> {
+    unsafe fn tree_hopper_copy(&self) -> Self {
+        ptr::read(self)
+    }
+}
+
+impl<BorrowType, K, V> NodeRef<BorrowType, K, V, marker::Leaf> {
+    pub fn try_ascend(
+        self,
+    ) -> Option<Handle<NodeRef<BorrowType, K, V, marker::Internal>, marker::Edge>> {
+        self.ascend().ok()
+    }
+}
+
+impl<BorrowType, K, V> NodeRef<BorrowType, K, V, marker::Internal> {
+    pub fn try_ascend(
+        self,
+    ) -> Option<Handle<NodeRef<BorrowType, K, V, marker::Internal>, marker::Edge>> {
+        self.ascend().ok()
+    }
+}
+
+macro_rules! def_tree_hopper {
+    { $next_kv:ident $next_edge:ident $initial_leaf_edge:ident $ascend:ident
+      $borrow_type:ty >> $($decl:tt)+} => {
+        /// Positions the given leaf edge handle on the next leaf edge
+        /// and returns the KV in between.
+        /// If there is no further KV, changes nothing and returns None.
+        /// Passing in the handle by value would be more elegant
+        /// (e.g. would eliminate the first tree_hopper_copy call),
+        /// but slows down iteration according to benchmarks.
```

Comment:
It's not really correct to say it changes nothing in the `deallocate_and_ascend` case -- in fact the original handle could be left dangling, no? I suppose that's currently avoided because the callers expect/know there is always another item, since they track their length.

I'm not really comfortable with quietly copying handles in the `Mut` and `Owned` cases. You did mark those `unsafe`, at least, but there's no indication to callers about the actual danger involved. Compare this to the `reborrow_mut` methods which say, "Beware, as this method is very dangerous, doubly so since it may not immediately appear dangerous."

I understand the original iterators had this effect too with their `ptr::read` duplication, but that was local, and it really should have been commented as well. You mention possibly passing in the handle by value, where I guess we could keep the responsibility for that copy on the caller's side -- how significant was your observed slowdown? It seems like that shouldn't impact inlined performance at all, just moving which side of the call does the copy.

Reply:
> It's not really correct to say it changes nothing in the `deallocate_and_ascend` case

Indeed, that's an awkward case.

> I suppose that's currently avoided because the callers expect/know there is always another item, since they track their length.

Yes, and in addition, the current callers (apart from the Drop implementation) are double-ended iterators. They could stop when front and back pointer are equal instead of tracking length, but they cannot use this "end-of-tree" return value. The only code actually using it is my drain_filter implementation, and there it's just a wee bit more elegant than tracking length. I mostly just liked the idea of having a totally safe function on immutable trees.

> there's no indication to callers about the actual danger involved.

I'm not really sure what the danger is. To me, it's that these hopping functions return two mutable handles pointing inside the same structure, so once the caller uses either to change something, the other one is on fire.

> You mention possibly passing in the handle by value, where I guess we could keep the responsibility for that copy on the caller's side -- how significant was your observed slowdown?

Between 5 and 10%. Normally within the range of noise, but so consistent over various changes back and forth that I guess it's really there. I [tried in vain to reproduce under laboratory conditions and asked around](https://users.rust-lang.org/t/why-are-mut-parameters-faster-than-moving-values/35176), but no real conclusion.

Reply:
> I'm not really sure what the danger is. To me, it's that these hopping functions return two mutable handles pointing inside the same structure, so once the caller uses either to change something, the other one is on fire.

Yes, invalidation like that is obviously bad. A simpler example is just the existence of safe `Handle::into_kv_mut` -- then separate handles could give you aliasing `&mut`, which is UB. We can admonish btree developers not to do that, but the point is that we've bypassed the compiler's usual protections against these hazards.

Reply:
> _not_ extracting the KV right away means we have to keep multiple handles alive, which is quietly dangerous in the `Mut` and `Owned` cases

I'm relatively new to Rust and don't grasp much of unsafety. I think I understand why extracting KV right away would be better: to reduce the attack surface of mistakes or abuse.

But the initiative to keep multiple handles alive in the `Mut` case was #58431, specifically done to make things, well, safer? better? possible? compatible with stacked borrows (just showing off I learnt a new term today). I don't really understand what the problem was, nor why it wasn't a problem in the `Owned` case, so definitely not how to extract KV right away in the `Mut` case and avoid #58431.